cloud/vsphere/utils.go

```diff
@@ -0,0 +1,83 @@
+package vsphere
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/labels"
+	vsphereconfig "sigs.k8s.io/cluster-api-provider-vsphere/cloud/vsphere/vsphereproviderconfig"
+	vsphereconfigv1 "sigs.k8s.io/cluster-api-provider-vsphere/cloud/vsphere/vsphereproviderconfig/v1alpha1"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	v1alpha1 "sigs.k8s.io/cluster-api/pkg/client/informers_generated/externalversions/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/util"
+)
+
+// GetMasterForCluster returns the master nodes for the given cluster
+func GetMasterForCluster(cluster *clusterv1.Cluster, lister v1alpha1.Interface) ([]*clusterv1.Machine, error) {
+	masters := make([]*clusterv1.Machine, 0)
+	machines, err := lister.Machines().Lister().Machines(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, machine := range machines {
+		//glog.Infof("%v", machine)
+		if util.IsMaster(machine) {
+			masters = append(masters, machine)
+			// Return the first master for now. Need to handle the multi-master case better
+			glog.Infof("Found the master VM %s", machine.Name)
+		}
+	}
+	if len(masters) == 0 {
+		glog.Infof("No master node found for the cluster %s", cluster.Name)
+	}
+	return masters, nil
+}
+
+func GetIP(_ *clusterv1.Cluster, machine *clusterv1.Machine) (string, error) {
+	if machine.ObjectMeta.Annotations != nil {
+		if ip, ok := machine.ObjectMeta.Annotations[VmIpAnnotationKey]; ok {
+			glog.Infof("Returning IP from machine annotation %s", ip)
+			return ip, nil
+		}
+	}
+	return "", errors.New("could not get IP")
+}
+
+func GetMachineProviderStatus(machine *clusterv1.Machine) (*vsphereconfig.VsphereMachineProviderStatus, error) {
+	if machine.Status.ProviderStatus != nil {
```

Comment:
hey, there is a Go idiom that basically says to return as soon as possible so that you don't form complex if-thens.  By checking for the valid case and creating a big if-then, you're going against the idiom.  I'd keep everything inside the if here, but structure this if-then as follows,

```
if machine.Status.ProviderStatus == nil {
  return nil, nil
}
```
...